torch/ao/quantization/backend_config/backend_config.py

```diff
@@ -0,0 +1,388 @@
+        return conf
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Convert this `BackendConfig` to a dictionary with the items described in
+        :func:`~torch.ao.quantization.backend_config.BackendConfig.from_dict`.
+        """
+        return {
+            NAME_DICT_KEY: self.name,
+            CONFIGS_DICT_KEY: [c.to_dict() for c in self.configs.values()],
+        }
+
+
+class BackendPatternConfig:
+    """
+    Config for ops defined in :class:`~torch.ao.quantization.backend_config.BackendConfig`.
+
+    The user can configure how a operator pattern graph is handled on a given backend using the following methods:
+        `set_observation_type`: sets how observers should be inserted for this op.
+            See :class:`~torch.ao.quantization.backend_config.ObservationType`
+        `add_dtype_config`: sets the supported data types for this op
+        `set_root_module`: sets the module that represents the root for this op
+        `set_qat_module`: sets the module that represents the QAT implementation for this op
+        `set_reference_quantized_module`: sets the module that represents the reference quantized
+            implementation for this op's root module.
+        `set_fused_module`: sets the module that represents the fused implementation for this op
+        `set_fuser_method`: sets the function that specifies how to fuse the pattern for this op
+
+    For a detailed example usage, see :class:`~torch.ao.quantization.backend_config.BackendConfig`.
+    """
+    def __init__(self, pattern: Pattern):
+        self.pattern = pattern
+        self.observation_type = ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT
+        self.dtype_configs: List[DTypeConfig] = []
+        self.root_module: Optional[torch.nn.Module] = None
+        self.qat_module: Optional[torch.nn.Module] = None
+        self.reference_quantized_module: Optional[torch.nn.Module] = None
+        self.fused_module: Optional[torch.nn.Module] = None
+        self.fuser_method: Optional[Callable] = None
+
+        # Temporary/internal configs
+        self._root_node_getter: Optional[Callable] = None
+        self._extra_inputs_getter: Optional[Callable] = None
+        self._num_tensor_args_to_observation_type: Dict[int, ObservationType] = {}
+        self._input_type_to_index: Dict[str, int] = {}
+        self._input_output_observed: Optional[bool] = None
+        self._overwrite_output_fake_quantize: Optional[_PartialWrapper] = None
+        self._overwrite_output_observer: Optional[_PartialWrapper] = None
+
+    def set_observation_type(self, observation_type: ObservationType) -> BackendPatternConfig:
+        """
+        Set how observers should be inserted for this op.
+        """
+        self.observation_type = observation_type
+        return self
+
+    def add_dtype_config(self, dtype_config: DTypeConfig) -> BackendPatternConfig:
+        """
+        Set the supported data types for this op.
```

Comment:
nit: comments need to be updated, also wondering if we want to give something more useful here, e.g. include some examples